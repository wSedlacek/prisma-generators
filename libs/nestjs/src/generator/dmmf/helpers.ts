// TODO: support not string args
export const modelAttributeRegex = /(@@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export const parseDocumentationAttributes = (
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: 'model' | 'field'
) => {
  const attributeRegex =
    expectedAttributeKind === 'model'
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return undefined;
  }

  return attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
};
