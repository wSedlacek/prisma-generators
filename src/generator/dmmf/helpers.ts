// TODO: support not string args
export const modelAttributeRegex = /(@@TypeGraphQL\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@TypeGraphQL\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export function parseDocumentationAttributes(
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: "model" | "field",
) {
  const attributeRegex =
    expectedAttributeKind === "model"
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return;
  }
  const attributeArgs = attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
  return attributeArgs;
}
