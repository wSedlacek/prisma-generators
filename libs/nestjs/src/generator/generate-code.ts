import { DMMF as PrismaDMMF } from '@prisma/client/runtime/dmmf-types';
import path from 'path';
import { CompilerOptions, ModuleKind, Project, ScriptTarget } from 'ts-morph';

import { isDefined } from '../utils';
import { generateArgsTypeClassFromArgs } from './args-class';
import {
  argsFolderName,
  crudResolversFolderName,
  enumsFolderName,
  inputsFolderName,
  modelsFolderName,
  outputsFolderName,
  relationsResolversFolderName,
  resolversFolderName,
} from './config';
import { DmmfDocument } from './dmmf/dmmf-document';
import { DMMF } from './dmmf/types';
import { generateEnumFromDef } from './enum';
import { noop } from './helpers';
import {
  generateArgsBarrelFile,
  generateEnumsBarrelFile,
  generateIndexFile,
  generateInputsBarrelFile,
  generateModelsBarrelFile,
  generateOutputsBarrelFile,
  generateResolversBarrelFile,
} from './imports';
import { generateObjectTypeClassFromModel } from './model-type-class';
import { GenerateCodeOptions } from './options';
import { generateCrudResolverClassFromMapping } from './resolvers/full-crud';
import { generateRelationsResolverClassesFromModel } from './resolvers/relations';
import { generateActionResolverClass } from './resolvers/separate-action';
import {
  generateInputTypeClassFromType,
  generateOutputTypeClassFromType,
} from './type-class';

const baseCompilerOptions: CompilerOptions = {
  target: ScriptTarget.ES2019,
  module: ModuleKind.CommonJS,
  emitDecoratorMetadata: true,
  experimentalDecorators: true,
};

export const generateCode = async (
  dmmf: PrismaDMMF.Document,
  options: GenerateCodeOptions,
  log: (msg: string) => void = noop
) => {
  const baseDirPath = options.outputDirPath;
  const emitTranspiledCode =
    options.emitTranspiledCode ??
    options.outputDirPath.includes('node_modules');
  const project = new Project({
    compilerOptions: {
      ...baseCompilerOptions,
      ...(emitTranspiledCode ? { declaration: true } : {}),
    },
  });
  const resolversDirPath = path.resolve(baseDirPath, resolversFolderName);
  const modelNames = dmmf.datamodel.models.map((model) => model.name);

  log('Transforming dmmfDocument...');
  const dmmfDocument = new DmmfDocument(dmmf, options);

  log('Generating enums...');
  const datamodelEnumNames = dmmfDocument.datamodel.enums.map(
    (enumDef) => enumDef.typeName
  );
  dmmfDocument.datamodel.enums.forEach((enumDef) =>
    generateEnumFromDef(project, baseDirPath, enumDef)
  );
  dmmfDocument.schema.enums
    // skip enums from datamodel
    .filter((enumDef) => !datamodelEnumNames.includes(enumDef.typeName))
    .forEach((enumDef) => generateEnumFromDef(project, baseDirPath, enumDef));
  const emittedEnumNames = [
    ...new Set([
      ...dmmfDocument.schema.enums.map((it) => it.typeName),
      ...dmmfDocument.datamodel.enums.map((it) => it.typeName),
    ]),
  ];
  const enumsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, enumsFolderName, 'index.ts'),
    undefined,
    { overwrite: true }
  );
  generateEnumsBarrelFile(enumsBarrelExportSourceFile, emittedEnumNames);

  log('Generating models...');
  dmmfDocument.datamodel.models.forEach((model) =>
    generateObjectTypeClassFromModel(project, baseDirPath, model, dmmfDocument)
  );
  const modelsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, modelsFolderName, 'index.ts'),
    undefined,
    { overwrite: true }
  );
  generateModelsBarrelFile(
    modelsBarrelExportSourceFile,
    dmmfDocument.datamodel.models.map((it) => it.typeName)
  );

  log('Generating output types...');
  const rootTypes = dmmfDocument.schema.outputTypes.filter((type) =>
    ['Query', 'Mutation'].includes(type.name)
  );
  const outputTypesToGenerate = dmmfDocument.schema.outputTypes.filter(
    // skip generating models and root resolvers
    (type) => !modelNames.includes(type.name) && !rootTypes.includes(type)
  );
  const outputTypesFieldsArgsToGenerate = outputTypesToGenerate
    .map((it) => it.fields)
    .reduce((a, b) => a.concat(b), [])
    .filter((it): it is Required<DMMF.OutputSchemaField> => !!it.argsTypeName);
  outputTypesToGenerate.forEach((type) =>
    generateOutputTypeClassFromType(
      project,
      resolversDirPath,
      type,
      dmmfDocument
    )
  );
  const outputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      outputsFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateOutputsBarrelFile(
    outputsBarrelExportSourceFile,
    outputTypesToGenerate.map((it) => it.typeName),
    outputTypesFieldsArgsToGenerate.length > 0
  );

  if (outputTypesFieldsArgsToGenerate.length > 0) {
    log('Generating output types args...');
    outputTypesFieldsArgsToGenerate.forEach((field) => {
      generateArgsTypeClassFromArgs(
        project,
        path.resolve(resolversDirPath, outputsFolderName),
        field.args,
        field.argsTypeName,
        dmmfDocument,
        2
      );
    });

    const outputsArgsBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        outputsFolderName,
        argsFolderName,
        'index.ts'
      ),
      undefined,
      { overwrite: true }
    );
    generateArgsBarrelFile(
      outputsArgsBarrelExportSourceFile,
      outputTypesFieldsArgsToGenerate
        .map((it) => it.argsTypeName)
        .filter(isDefined)
    );
  }

  log('Generating input types...');

  dmmfDocument.schema.inputTypes.forEach((type) =>
    generateInputTypeClassFromType(
      project,
      resolversDirPath,
      type,
      dmmfDocument
    )
  );

  const inputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      inputsFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateInputsBarrelFile(
    inputsBarrelExportSourceFile,
    dmmfDocument.schema.inputTypes.map((it) => it.typeName)
  );

  if (dmmfDocument.relationModels.length > 0) {
    log('Generating relation resolvers...');

    dmmfDocument.relationModels.forEach((relationModel) =>
      generateRelationsResolverClassesFromModel(
        project,
        baseDirPath,
        dmmfDocument,
        relationModel
      )
    );

    const relationResolversBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        'index.ts'
      ),
      undefined,
      { overwrite: true }
    );
    generateResolversBarrelFile(
      'relations',
      relationResolversBarrelExportSourceFile,
      dmmfDocument.relationModels.map((relationModel) => ({
        resolverName: relationModel.resolverName,
        modelName: relationModel.model.typeName,
        hasSomeArgs: relationModel.relationFields.some(
          (field) => field.argsTypeName !== undefined
        ),
      }))
    );

    log('Generating relation resolver args...');
    dmmfDocument.relationModels.forEach((relationModelData) => {
      const resolverDirPath = path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        relationModelData.model.typeName
      );

      relationModelData.relationFields
        .filter((field) => field.argsTypeName)
        .forEach((field) => {
          generateArgsTypeClassFromArgs(
            project,
            resolverDirPath,
            field.outputTypeField.args,
            field.argsTypeName as string,
            dmmfDocument
          );
        });

      const argTypeNames = relationModelData.relationFields
        .filter((it) => it.argsTypeName !== undefined)
        .map((it) => it.argsTypeName)
        .filter(isDefined);

      if (argTypeNames.length) {
        const barrelExportSourceFile = project.createSourceFile(
          path.resolve(resolverDirPath, argsFolderName, 'index.ts'),
          undefined,
          { overwrite: true }
        );
        generateArgsBarrelFile(barrelExportSourceFile, argTypeNames);
      }
    });
  }

  log('Generating crud resolvers...');
  dmmfDocument.mappings.forEach((mapping) => {
    const model = dmmfDocument.datamodel.models.find(
      (it) => it.name === mapping.model
    ) as DMMF.Model;
    generateCrudResolverClassFromMapping(
      project,
      baseDirPath,
      mapping,
      model,
      dmmfDocument
    );
    mapping.actions.forEach((action) => {
      const actionModel = dmmfDocument.datamodel.models.find(
        (it) => it.name === mapping.model
      ) as DMMF.Model;
      generateActionResolverClass(
        project,
        baseDirPath,
        actionModel,
        action,
        mapping,
        dmmfDocument
      );
    });
  });
  const crudResolversBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      crudResolversFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateResolversBarrelFile(
    'crud',
    crudResolversBarrelExportSourceFile,
    dmmfDocument.mappings.map((mapping) => {
      const model = dmmfDocument.datamodel.models.find(
        (it) => it.name === mapping.model
      ) as DMMF.Model;

      return {
        modelName: model.typeName,
        resolverName: mapping.resolverName,
        actionResolverNames: mapping.actions.map((it) => it.actionResolverName),
        hasSomeArgs: mapping.actions.some(
          (action) => action.argsTypeName !== undefined
        ),
      };
    })
  );

  log('Generating crud resolvers args...');
  dmmfDocument.mappings.forEach((mapping) => {
    const actionsWithArgs = mapping.actions.filter(
      (it) => it.argsTypeName !== undefined
    );

    if (actionsWithArgs.length) {
      const model = dmmfDocument.datamodel.models.find(
        (it) => it.name === mapping.model
      ) as DMMF.Model;
      const resolverDirPath = path.resolve(
        baseDirPath,
        resolversFolderName,
        crudResolversFolderName,
        model.typeName
      );
      actionsWithArgs.forEach((action) => {
        generateArgsTypeClassFromArgs(
          project,
          resolverDirPath,
          action.method.args,
          action.argsTypeName as string,
          dmmfDocument
        );
      });
      const barrelExportSourceFile = project.createSourceFile(
        path.resolve(resolverDirPath, argsFolderName, 'index.ts'),
        undefined,
        { overwrite: true }
      );
      generateArgsBarrelFile(
        barrelExportSourceFile,
        actionsWithArgs.map((it) => it.argsTypeName).filter(isDefined)
      );
    }
  });

  log('Generating index file');
  const indexSourceFile = project.createSourceFile(
    `${baseDirPath}/index.ts`,
    undefined,
    { overwrite: true }
  );
  generateIndexFile(indexSourceFile, dmmfDocument.relationModels.length > 0);

  log('Emitting generated code files');
  if (emitTranspiledCode) {
    await project.emit();
  } else {
    for (const file of project.getSourceFiles()) {
      file.formatText({ indentSize: 2 });
    }
    await project.save();
  }
};
