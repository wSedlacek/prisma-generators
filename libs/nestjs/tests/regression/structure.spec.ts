import directoryTree from 'directory-tree';
import { promises as fs } from 'fs';

import { dedent } from 'ts-dedent';
import { generateArtifactsDirPath } from '../helpers/artifacts-dir';
import { generateCodeFromSchema } from '../helpers/generate-code';
import { prisma } from '../helpers/prisma-template';
import { stringifyDirectoryTrees } from '../helpers/structure';

describe('structure', () => {
  let outputDirPath: string;
  let complexDatamodel: string;

  beforeEach(async () => {
    outputDirPath = generateArtifactsDirPath('regression-structure');
    await fs.mkdir(outputDirPath, { recursive: true });

    complexDatamodel = prisma`
      enum Color {
        RED
        GREEN
        BLUE
      }

      model User {
        id    Int      @id @default(autoincrement())
        name  String?
        posts Post[]
      }

      model Post {
        uuid      String  @id @default(cuid())
        content   String
        author    User    @relation(fields: [authorId], references: [id])
        authorId  Int
        color     Color
      }
  `;
  });

  it('should generate proper folders structure and file names for complex datamodel', async () => {
    await generateCodeFromSchema(complexDatamodel, { outputDirPath });
    const directoryStructure = directoryTree(outputDirPath);
    const directoryStructureString = dedent`\n[type-graphql]
    ${stringifyDirectoryTrees(directoryStructure.children, 2)}`;

    // FIXME: replace with `.toMatchInlineSnapshot()` when it starts working again
    expect(directoryStructureString).toMatchSnapshot('structure');
  });

  it('should generate *.js and *.d.ts files when emitTranspiledCode is set to true', async () => {
    await generateCodeFromSchema(complexDatamodel, {
      outputDirPath,
      emitTranspiledCode: true,
    });
    const directoryStructure = directoryTree(outputDirPath);
    const directoryStructureString = `\n[type-graphql]\n${stringifyDirectoryTrees(
      directoryStructure.children,
      2
    )}`;

    // FIXME: replace with `.toMatchInlineSnapshot()` when it starts working again
    expect(directoryStructureString).toMatchSnapshot('structure');
  }, 20000);

  it('should generate proper folders and file names when model is renamed', async () => {
    const schema = prisma`
      enum Color {
        RED
        GREEN
        BLUE
      }

      /// @@NestJS.type("RenamedUser")
      model User {
        id    Int      @id @default(autoincrement())
        name  String?
        posts Post[]
      }

      /// @@NestJS.type("RenamedPost")
      model Post {
        uuid      String  @id @default(cuid())
        content   String
        author    User    @relation(fields: [authorId], references: [id])
        authorId  Int
        color     Color
      }
    `;

    await generateCodeFromSchema(schema, { outputDirPath });
    const directoryStructure = directoryTree(outputDirPath);
    const directoryStructureString = dedent`\n[type-graphql]
      ${stringifyDirectoryTrees(directoryStructure.children, 2)}`;

    // FIXME: replace with `.toMatchInlineSnapshot()` when it starts working again
    expect(directoryStructureString).toMatchSnapshot('structure');
  });
});
