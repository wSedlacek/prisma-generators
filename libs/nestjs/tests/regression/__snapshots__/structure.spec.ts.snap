// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`structure should generate *.js and *.d.ts files when emitTranspiledCode is set to true: structure 1`] = `
"
[type-graphql]
  [enums]
    Color.d.ts
    Color.js
    PostDistinctFieldEnum.d.ts
    PostDistinctFieldEnum.js
    SortOrder.d.ts
    SortOrder.js
    UserDistinctFieldEnum.d.ts
    UserDistinctFieldEnum.js
    index.d.ts
    index.js
  index.d.ts
  index.js
  [models]
    Post.d.ts
    Post.js
    User.d.ts
    User.js
    index.d.ts
    index.js
  [resolvers]
    [crud]
      [Post]
        AggregatePostResolver.d.ts
        AggregatePostResolver.js
        CreatePostResolver.d.ts
        CreatePostResolver.js
        DeleteManyPostResolver.d.ts
        DeleteManyPostResolver.js
        DeletePostResolver.d.ts
        DeletePostResolver.js
        FindManyPostResolver.d.ts
        FindManyPostResolver.js
        FindOnePostResolver.d.ts
        FindOnePostResolver.js
        PostCrudResolver.d.ts
        PostCrudResolver.js
        UpdateManyPostResolver.d.ts
        UpdateManyPostResolver.js
        UpdatePostResolver.d.ts
        UpdatePostResolver.js
        UpsertPostResolver.d.ts
        UpsertPostResolver.js
        [args]
          AggregatePostArgs.d.ts
          AggregatePostArgs.js
          CreatePostArgs.d.ts
          CreatePostArgs.js
          DeleteManyPostArgs.d.ts
          DeleteManyPostArgs.js
          DeletePostArgs.d.ts
          DeletePostArgs.js
          FindManyPostArgs.d.ts
          FindManyPostArgs.js
          FindOnePostArgs.d.ts
          FindOnePostArgs.js
          UpdateManyPostArgs.d.ts
          UpdateManyPostArgs.js
          UpdatePostArgs.d.ts
          UpdatePostArgs.js
          UpsertPostArgs.d.ts
          UpsertPostArgs.js
          index.d.ts
          index.js
      [User]
        AggregateUserResolver.d.ts
        AggregateUserResolver.js
        CreateUserResolver.d.ts
        CreateUserResolver.js
        DeleteManyUserResolver.d.ts
        DeleteManyUserResolver.js
        DeleteUserResolver.d.ts
        DeleteUserResolver.js
        FindManyUserResolver.d.ts
        FindManyUserResolver.js
        FindOneUserResolver.d.ts
        FindOneUserResolver.js
        UpdateManyUserResolver.d.ts
        UpdateManyUserResolver.js
        UpdateUserResolver.d.ts
        UpdateUserResolver.js
        UpsertUserResolver.d.ts
        UpsertUserResolver.js
        UserCrudResolver.d.ts
        UserCrudResolver.js
        [args]
          AggregateUserArgs.d.ts
          AggregateUserArgs.js
          CreateUserArgs.d.ts
          CreateUserArgs.js
          DeleteManyUserArgs.d.ts
          DeleteManyUserArgs.js
          DeleteUserArgs.d.ts
          DeleteUserArgs.js
          FindManyUserArgs.d.ts
          FindManyUserArgs.js
          FindOneUserArgs.d.ts
          FindOneUserArgs.js
          UpdateManyUserArgs.d.ts
          UpdateManyUserArgs.js
          UpdateUserArgs.d.ts
          UpdateUserArgs.js
          UpsertUserArgs.d.ts
          UpsertUserArgs.js
          index.d.ts
          index.js
      index.d.ts
      index.js
    [inputs]
      EnumColorFieldUpdateOperationsInput.d.ts
      EnumColorFieldUpdateOperationsInput.js
      EnumColorFilter.d.ts
      EnumColorFilter.js
      IntFilter.d.ts
      IntFilter.js
      NestedEnumColorFilter.d.ts
      NestedEnumColorFilter.js
      NestedIntFilter.d.ts
      NestedIntFilter.js
      NestedStringFilter.d.ts
      NestedStringFilter.js
      NestedStringNullableFilter.d.ts
      NestedStringNullableFilter.js
      NullableStringFieldUpdateOperationsInput.d.ts
      NullableStringFieldUpdateOperationsInput.js
      PostCreateInput.d.ts
      PostCreateInput.js
      PostCreateManyWithoutAuthorInput.d.ts
      PostCreateManyWithoutAuthorInput.js
      PostCreateWithoutAuthorInput.d.ts
      PostCreateWithoutAuthorInput.js
      PostListRelationFilter.d.ts
      PostListRelationFilter.js
      PostOrderByInput.d.ts
      PostOrderByInput.js
      PostScalarWhereInput.d.ts
      PostScalarWhereInput.js
      PostUpdateInput.d.ts
      PostUpdateInput.js
      PostUpdateManyDataInput.d.ts
      PostUpdateManyDataInput.js
      PostUpdateManyMutationInput.d.ts
      PostUpdateManyMutationInput.js
      PostUpdateManyWithWhereNestedInput.d.ts
      PostUpdateManyWithWhereNestedInput.js
      PostUpdateManyWithoutAuthorInput.d.ts
      PostUpdateManyWithoutAuthorInput.js
      PostUpdateWithWhereUniqueWithoutAuthorInput.d.ts
      PostUpdateWithWhereUniqueWithoutAuthorInput.js
      PostUpdateWithoutAuthorDataInput.d.ts
      PostUpdateWithoutAuthorDataInput.js
      PostUpsertWithWhereUniqueWithoutAuthorInput.d.ts
      PostUpsertWithWhereUniqueWithoutAuthorInput.js
      PostWhereInput.d.ts
      PostWhereInput.js
      PostWhereUniqueInput.d.ts
      PostWhereUniqueInput.js
      StringFieldUpdateOperationsInput.d.ts
      StringFieldUpdateOperationsInput.js
      StringFilter.d.ts
      StringFilter.js
      StringNullableFilter.d.ts
      StringNullableFilter.js
      UserCreateInput.d.ts
      UserCreateInput.js
      UserCreateOneWithoutPostsInput.d.ts
      UserCreateOneWithoutPostsInput.js
      UserCreateWithoutPostsInput.d.ts
      UserCreateWithoutPostsInput.js
      UserOrderByInput.d.ts
      UserOrderByInput.js
      UserRelationFilter.d.ts
      UserRelationFilter.js
      UserUpdateInput.d.ts
      UserUpdateInput.js
      UserUpdateManyMutationInput.d.ts
      UserUpdateManyMutationInput.js
      UserUpdateOneRequiredWithoutPostsInput.d.ts
      UserUpdateOneRequiredWithoutPostsInput.js
      UserUpdateWithoutPostsDataInput.d.ts
      UserUpdateWithoutPostsDataInput.js
      UserUpsertWithoutPostsInput.d.ts
      UserUpsertWithoutPostsInput.js
      UserWhereInput.d.ts
      UserWhereInput.js
      UserWhereUniqueInput.d.ts
      UserWhereUniqueInput.js
      index.d.ts
      index.js
    [outputs]
      AggregatePost.d.ts
      AggregatePost.js
      AggregateUser.d.ts
      AggregateUser.js
      BatchPayload.d.ts
      BatchPayload.js
      PostAvgAggregateOutputType.d.ts
      PostAvgAggregateOutputType.js
      PostMaxAggregateOutputType.d.ts
      PostMaxAggregateOutputType.js
      PostMinAggregateOutputType.d.ts
      PostMinAggregateOutputType.js
      PostSumAggregateOutputType.d.ts
      PostSumAggregateOutputType.js
      UserAvgAggregateOutputType.d.ts
      UserAvgAggregateOutputType.js
      UserMaxAggregateOutputType.d.ts
      UserMaxAggregateOutputType.js
      UserMinAggregateOutputType.d.ts
      UserMinAggregateOutputType.js
      UserSumAggregateOutputType.d.ts
      UserSumAggregateOutputType.js
      index.d.ts
      index.js
    [relations]
      [Post]
        PostRelationsResolver.d.ts
        PostRelationsResolver.js
      [User]
        UserRelationsResolver.d.ts
        UserRelationsResolver.js
        [args]
          UserPostsArgs.d.ts
          UserPostsArgs.js
          index.d.ts
          index.js
      index.d.ts
      index.js
"
`;

exports[`structure should generate proper folders and file names when model is renamed: structure 1`] = `
"\\\\n[type-graphql]
  [enums]
    Color.ts
    RenamedPostDistinctFieldEnum.ts
    RenamedUserDistinctFieldEnum.ts
    SortOrder.ts
    index.ts
  index.ts
  [models]
    RenamedPost.ts
    RenamedUser.ts
    index.ts
  [resolvers]
    [crud]
      [RenamedPost]
        AggregateRenamedPostResolver.ts
        CreateRenamedPostResolver.ts
        DeleteManyRenamedPostResolver.ts
        DeleteRenamedPostResolver.ts
        FindManyRenamedPostResolver.ts
        FindOneRenamedPostResolver.ts
        RenamedPostCrudResolver.ts
        UpdateManyRenamedPostResolver.ts
        UpdateRenamedPostResolver.ts
        UpsertRenamedPostResolver.ts
        [args]
          AggregateRenamedPostArgs.ts
          CreateRenamedPostArgs.ts
          DeleteManyRenamedPostArgs.ts
          DeleteRenamedPostArgs.ts
          FindManyRenamedPostArgs.ts
          FindOneRenamedPostArgs.ts
          UpdateManyRenamedPostArgs.ts
          UpdateRenamedPostArgs.ts
          UpsertRenamedPostArgs.ts
          index.ts
      [RenamedUser]
        AggregateRenamedUserResolver.ts
        CreateRenamedUserResolver.ts
        DeleteManyRenamedUserResolver.ts
        DeleteRenamedUserResolver.ts
        FindManyRenamedUserResolver.ts
        FindOneRenamedUserResolver.ts
        RenamedUserCrudResolver.ts
        UpdateManyRenamedUserResolver.ts
        UpdateRenamedUserResolver.ts
        UpsertRenamedUserResolver.ts
        [args]
          AggregateRenamedUserArgs.ts
          CreateRenamedUserArgs.ts
          DeleteManyRenamedUserArgs.ts
          DeleteRenamedUserArgs.ts
          FindManyRenamedUserArgs.ts
          FindOneRenamedUserArgs.ts
          UpdateManyRenamedUserArgs.ts
          UpdateRenamedUserArgs.ts
          UpsertRenamedUserArgs.ts
          index.ts
      index.ts
    [inputs]
      EnumColorFieldUpdateOperationsInput.ts
      EnumColorFilter.ts
      IntFilter.ts
      NestedEnumColorFilter.ts
      NestedIntFilter.ts
      NestedStringFilter.ts
      NestedStringNullableFilter.ts
      NullableStringFieldUpdateOperationsInput.ts
      RenamedPostCreateInput.ts
      RenamedPostCreateManyWithoutAuthorInput.ts
      RenamedPostCreateWithoutAuthorInput.ts
      RenamedPostListRelationFilter.ts
      RenamedPostOrderByInput.ts
      RenamedPostScalarWhereInput.ts
      RenamedPostUpdateInput.ts
      RenamedPostUpdateManyDataInput.ts
      RenamedPostUpdateManyMutationInput.ts
      RenamedPostUpdateManyWithWhereNestedInput.ts
      RenamedPostUpdateManyWithoutAuthorInput.ts
      RenamedPostUpdateWithWhereUniqueWithoutAuthorInput.ts
      RenamedPostUpdateWithoutAuthorDataInput.ts
      RenamedPostUpsertWithWhereUniqueWithoutAuthorInput.ts
      RenamedPostWhereInput.ts
      RenamedPostWhereUniqueInput.ts
      RenamedUserCreateInput.ts
      RenamedUserCreateOneWithoutPostsInput.ts
      RenamedUserCreateWithoutPostsInput.ts
      RenamedUserOrderByInput.ts
      RenamedUserRelationFilter.ts
      RenamedUserUpdateInput.ts
      RenamedUserUpdateManyMutationInput.ts
      RenamedUserUpdateOneRequiredWithoutPostsInput.ts
      RenamedUserUpdateWithoutPostsDataInput.ts
      RenamedUserUpsertWithoutPostsInput.ts
      RenamedUserWhereInput.ts
      RenamedUserWhereUniqueInput.ts
      StringFieldUpdateOperationsInput.ts
      StringFilter.ts
      StringNullableFilter.ts
      index.ts
    [outputs]
      AggregateRenamedPost.ts
      AggregateRenamedUser.ts
      BatchPayload.ts
      RenamedPostAvgAggregateOutputType.ts
      RenamedPostMaxAggregateOutputType.ts
      RenamedPostMinAggregateOutputType.ts
      RenamedPostSumAggregateOutputType.ts
      RenamedUserAvgAggregateOutputType.ts
      RenamedUserMaxAggregateOutputType.ts
      RenamedUserMinAggregateOutputType.ts
      RenamedUserSumAggregateOutputType.ts
      index.ts
    [relations]
      [RenamedPost]
        RenamedPostRelationsResolver.ts
      [RenamedUser]
        RenamedUserRelationsResolver.ts
        [args]
          RenamedUserPostsArgs.ts
          index.ts
      index.ts
"
`;

exports[`structure should generate proper folders structure and file names for complex datamodel: structure 1`] = `
"\\\\n[type-graphql]
  [enums]
    Color.ts
    PostDistinctFieldEnum.ts
    SortOrder.ts
    UserDistinctFieldEnum.ts
    index.ts
  index.ts
  [models]
    Post.ts
    User.ts
    index.ts
  [resolvers]
    [crud]
      [Post]
        AggregatePostResolver.ts
        CreatePostResolver.ts
        DeleteManyPostResolver.ts
        DeletePostResolver.ts
        FindManyPostResolver.ts
        FindOnePostResolver.ts
        PostCrudResolver.ts
        UpdateManyPostResolver.ts
        UpdatePostResolver.ts
        UpsertPostResolver.ts
        [args]
          AggregatePostArgs.ts
          CreatePostArgs.ts
          DeleteManyPostArgs.ts
          DeletePostArgs.ts
          FindManyPostArgs.ts
          FindOnePostArgs.ts
          UpdateManyPostArgs.ts
          UpdatePostArgs.ts
          UpsertPostArgs.ts
          index.ts
      [User]
        AggregateUserResolver.ts
        CreateUserResolver.ts
        DeleteManyUserResolver.ts
        DeleteUserResolver.ts
        FindManyUserResolver.ts
        FindOneUserResolver.ts
        UpdateManyUserResolver.ts
        UpdateUserResolver.ts
        UpsertUserResolver.ts
        UserCrudResolver.ts
        [args]
          AggregateUserArgs.ts
          CreateUserArgs.ts
          DeleteManyUserArgs.ts
          DeleteUserArgs.ts
          FindManyUserArgs.ts
          FindOneUserArgs.ts
          UpdateManyUserArgs.ts
          UpdateUserArgs.ts
          UpsertUserArgs.ts
          index.ts
      index.ts
    [inputs]
      EnumColorFieldUpdateOperationsInput.ts
      EnumColorFilter.ts
      IntFilter.ts
      NestedEnumColorFilter.ts
      NestedIntFilter.ts
      NestedStringFilter.ts
      NestedStringNullableFilter.ts
      NullableStringFieldUpdateOperationsInput.ts
      PostCreateInput.ts
      PostCreateManyWithoutAuthorInput.ts
      PostCreateWithoutAuthorInput.ts
      PostListRelationFilter.ts
      PostOrderByInput.ts
      PostScalarWhereInput.ts
      PostUpdateInput.ts
      PostUpdateManyDataInput.ts
      PostUpdateManyMutationInput.ts
      PostUpdateManyWithWhereNestedInput.ts
      PostUpdateManyWithoutAuthorInput.ts
      PostUpdateWithWhereUniqueWithoutAuthorInput.ts
      PostUpdateWithoutAuthorDataInput.ts
      PostUpsertWithWhereUniqueWithoutAuthorInput.ts
      PostWhereInput.ts
      PostWhereUniqueInput.ts
      StringFieldUpdateOperationsInput.ts
      StringFilter.ts
      StringNullableFilter.ts
      UserCreateInput.ts
      UserCreateOneWithoutPostsInput.ts
      UserCreateWithoutPostsInput.ts
      UserOrderByInput.ts
      UserRelationFilter.ts
      UserUpdateInput.ts
      UserUpdateManyMutationInput.ts
      UserUpdateOneRequiredWithoutPostsInput.ts
      UserUpdateWithoutPostsDataInput.ts
      UserUpsertWithoutPostsInput.ts
      UserWhereInput.ts
      UserWhereUniqueInput.ts
      index.ts
    [outputs]
      AggregatePost.ts
      AggregateUser.ts
      BatchPayload.ts
      PostAvgAggregateOutputType.ts
      PostMaxAggregateOutputType.ts
      PostMinAggregateOutputType.ts
      PostSumAggregateOutputType.ts
      UserAvgAggregateOutputType.ts
      UserMaxAggregateOutputType.ts
      UserMinAggregateOutputType.ts
      UserSumAggregateOutputType.ts
      index.ts
    [relations]
      [Post]
        PostRelationsResolver.ts
      [User]
        UserRelationsResolver.ts
        [args]
          UserPostsArgs.ts
          index.ts
      index.ts
"
`;
